File: C#/XamarinStore.iOS/Views/UrlUIImage.cs
using System;
using MonoTouch.UIKit;

namespace XamarinStore
{
	public class UrlUIImage : UIImageView
	{
		UIActivityIndicatorView progress;

		public UrlUIImage ()
		{
			AddSubview (progress = new UIActivityIndicatorView (UIActivityIndicatorViewStyle.WhiteLarge));
		}

		string url;

		public string Url {
			get {
				return url;
			}
			set {
				url = value;
				DownloadImage ();
			}
		}

		async void DownloadImage ()
		{
			if (string.IsNullOrEmpty (url))
				return;
			var t = FileCache.Download (url);
			if (t.IsCompleted) {
				Image = UIImage.FromFile(t.Result);
				return;
			}
			progress.StartAnimating ();
			var image = UIImage.FromFile(await t);

			UIView.Animate (.3, 
				() => Image = image,
				() => progress.StopAnimating ());

		}
		public override void LayoutSubviews ()
		{
			base.LayoutSubviews ();
			progress.Center = Center;
		}
	}
}

File: F#/XamarinStore.iOS/Views/UrlUIImage.fs
namespace XamarinStore

open System
open MonoTouch.UIKit

type UrlUIImage() as this =
    inherit UIImageView()

    let mutable (url:string) = null
    let progress = new UIActivityIndicatorView (UIActivityIndicatorViewStyle.WhiteLarge)

    do this.AddSubview progress

    member this.Url
        with get () = url
        and set value = url <- value
                        this.DownloadImage () |> ignore

    member this.DownloadImage () = async {
        if not (String.IsNullOrEmpty(url)) then
            let! t = FileCache.Download (url)
            progress.StartAnimating ()
            let image = UIImage.FromFile(t)

            UIView.Animate (0.3, 
                (fun () -> this.Image <- image),
                fun () -> progress.StopAnimating ()) }

    override this.LayoutSubviews () =
        base.LayoutSubviews ()
        progress.Center <- this.Center